ssz/sedes/base.py

```diff
@@ -104,77 +83,126 @@ def __init__(self, size: int):
 
-        if self.is_static_sized:
-            return content
+        section_0_size = _compute_section_0_size(element_sedes)
+
+        section_1_components = tuple(
+            sedes.serialize(item)
+            for item, sedes
+            in pairs
+            if not sedes.is_fixed_sized
+        )
+
+        if section_1_components:
+            offsets = tuple(accumulate(
+                operator.add,
+                map(len, section_1_components[:-1]),
+                section_0_size,
+            ))
         else:
-            validate_content_size(content)
-            size_prefix = get_size_prefix(content)
-            return size_prefix + content
+            offsets = ()
+
+        offsets_iter = iter(offsets)
+
+        section_0_components = tuple(
+            sedes.serialize(item) if sedes.is_fixed_sized else encode_offset(next(offsets_iter))
+            for item, sedes
+            in pairs
+        )
+
+        try:
+            next(offsets_iter)
+        except StopIteration:
+            pass
+        else:
+            raise DeserializationError("Did not consume all offsets while decoding value")
+
+        return b"".join(concatv(
+            section_0_components,
+            section_1_components,
+        ))
+
+    def deserialize(self, data: bytes) -> TCanonical:
+        stream = io.BytesIO(data)
+        value = self.s_deserialize(stream)
+        extra_data = stream.read()
+        if extra_data:
+            raise DeserializationError(f"Got {len(extra_data)} superfluous bytes")
+        return value
 
     @abstractmethod
-    def serialize_content(self, value: TSerializable) -> bytes:
+    def s_deserialize(self, stream: IO[bytes]) -> TCanonical:
         pass
 
+
+TSerializableElement = TypeVar("TSerializableElement")
+TDeserializedElement = TypeVar("TDeserializedElement")
+
+
+class HomogenousSequence(CompositeSedes[Sequence[Any], Tuple[Any]]):
+    element_sedes: TSedes
+
     #
-    # Deserialization
+    # Serialization
     #
-    def deserialize_segment(self, data: bytes, start_index: int) -> Tuple[TDeserialized, int]:
-        if self.is_static_sized:
-            content_size = self.get_static_size()
-            content, continuation_index = self.consume_bytes(data, start_index, content_size)
-        else:
-            prefix, content_start_index = self.consume_bytes(data, start_index, SIZE_PREFIX_SIZE)
-            length = int.from_bytes(prefix, "little")
-            content, continuation_index = self.consume_bytes(data, content_start_index, length)
-        return self.deserialize_content(content), continuation_index
+    def _iter_sedes(self, value: Sequence[Any]) -> Tuple[Tuple[Any, TSedes], ...]:
+        return tuple(zip(value, itertools.repeat(self.element_sedes)))
 
-    @abstractmethod
-    def deserialize_content(self, content: bytes) -> TDeserialized:
-        pass
+    #
+    # Tree hashing
+    #
+    def hash_tree_root(self, value: Sequence[Any]) -> bytes:
```

Comment:
It looks like this is only suitable for `Vector`, but not for `List` and it would make more sense to have this in `Vector` directly.